haystack/components/agents/agent.py

```diff
@@ -266,3 +247,115 @@ def run(
+        **kwargs: Dict[str, Any],
+    ) -> Dict[str, Any]:
+        """
+        Asynchronously process messages and execute tools until the exit condition is met.
+
+        This is the asynchronous version of the `run` method. It follows the same logic but uses
+        asynchronous operations where possible, such as calling the `run_async` method of the ChatGenerator
+        if available.
+
+        :param messages: List of chat messages to process
+        :param streaming_callback: A callback that will be invoked when a response is streamed from the LLM.
+        :param kwargs: Additional data to pass to the State schema used by the Agent.
+            The keys must match the schema defined in the Agent's `state_schema`.
+        :return: Dictionary containing messages and outputs matching the defined output types
+        """
+        if not self._is_warmed_up and hasattr(self.chat_generator, "warm_up"):
+            raise RuntimeError("The component Agent wasn't warmed up. Run 'warm_up()' before calling 'run_async()'.")
+
+        state = State(schema=self.state_schema, data=kwargs)
+
+        if self.system_prompt is not None:
+            messages = [ChatMessage.from_system(self.system_prompt)] + messages
+        state.set("messages", messages)
+
+        generator_inputs: Dict[str, Any] = {"tools": self.tools}
+
+        selected_callback = streaming_callback or self.streaming_callback
+        if selected_callback is not None:
+            generator_inputs["streaming_callback"] = selected_callback
+
+        # Repeat until the exit condition is met
+        counter = 0
+        while counter < self.max_agent_steps:
+            # 1. Call the ChatGenerator
+            # Check if the chat generator supports async execution
+            if getattr(self.chat_generator, "__haystack_supports_async__", False):
+                result = await self.chat_generator.run_async(messages=messages, **generator_inputs)
+                llm_messages = result["replies"]
+            else:
+                # Fall back to synchronous run if async is not available
+                loop = asyncio.get_running_loop()
+                result = await loop.run_in_executor(
+                    None, lambda: self.chat_generator.run(messages=messages, **generator_inputs)
+                )
+                llm_messages = result["replies"]
+
+            state.set("messages", llm_messages)
+
+            # 2. Check if any of the LLM responses contain a tool call
+            if not any(msg.tool_call for msg in llm_messages):
+                return {**state.data}
+
+            # 3. Call the ToolInvoker
+            # We only send the messages from the LLM to the tool invoker
+            # Check if the ToolInvoker supports async execution. Currently, it doesn't.
+            if getattr(self._tool_invoker, "__haystack_supports_async__", False):
+                tool_invoker_result = await self._tool_invoker.run_async(messages=llm_messages, state=state)
```

Comment:
Same as above, we may want to silence this mypy warning with `# type: ignore[attr-defined]`.